src/Paket.Core/Versioning/SemVer.fs

```diff
@@ -177,55 +255,59 @@ module SemVer =
                     | true, s when s < 0 -> failwith "no negatives!" 
-                    | _ -> ignore ()
+                    | _ -> ignore ()  // non-numeric parts are valid
 
                 if version.Contains("!") then 
                     failwithf "Invalid character found in %s" version
                 if version.Contains("..") then 
                     failwithf "Empty version part found in %s" version
-        
-                let firstDash = version.IndexOf("-")
+
                 let plusIndex = version.IndexOf("+")
 
-                let majorMinorPatch =
-                    let firstSigil = if firstDash > 0 then firstDash else plusIndex
-                    match firstSigil with
-                    | -1 -> version
+                let versionStr = 
+                    match plusIndex with
+                    | n when n < 0 -> version
                     | n -> version.Substring(0, n)
 
-                let prerelease = 
-                    match firstDash, plusIndex with
-                    | -1, _ -> ""
-                    | d, p when p = -1 -> version.Substring(d+1)
-                    | d, p -> version.Substring(d+1, (p - 1 - d) )
-            
-                /// there can only be one piece of build metadata, and it is signified by a + and then any number of dot-separated alpha-numeric groups.
-                /// this just greedily takes the whole remaining string :(
+                /// there can only be one piece of build metadata, and it is signified by + sign
+                /// and then any number of dot-separated alpha-numeric groups.
                 let buildmeta =
                     match plusIndex with
                     | -1 -> ""
-                    | n when plusIndex = version.Length - 1 -> ""
-                    | n -> version.Substring(plusIndex + 1)
-        
-                let major, minor, patch, build =
-                    match majorMinorPatch.Split([|'.'|]) with
-                    | [|M; m; p; b|] -> uint32 M, uint32 m, uint32 p, b
-                    | [|M; m; p; |] -> uint32 M, uint32 m, uint32 p, "0"
-                    | [|M; m;|] -> uint32 M, uint32 m, 0u, "0"
-                    | [|M;|] -> uint32 M, 0u, 0u, "0"
-                    | _ -> 0u, 0u, 0u, "0"
-
-                { Major = major
-                  Minor = minor
-                  Patch = patch
-                  Build = build
+                    | n when n = version.Length - 1 -> ""
+                    | n -> 
+                        let content = validContent.Match(version.Substring(n + 1))
+                        if content.Success then content.Value else ""
+
+                let fragments = expand [ '.'; '-' ] versionStr
+                /// matches over list of the version fragments *and* delimiters
+                let major, minor, patch, revision, suffix =
+                    match fragments with
+                    | (Int M)::"."::(Int m)::"."::(Int p)::"."::(Big b)::tail -> M, m, p, b, tail
```

Comment:
match including delimiters allows to parse as much of valid version parts as possible, and processing the rest as prerelease segments